File: server/src/schemas/student.py
"""Student-related schemas."""

from uuid import UUID

from pydantic import BaseModel


class StudentResponse(BaseModel):
    """Response model for student data."""

    student_tz: str
    student_name: str
    class_id: UUID | None
    class_name: str
    grade_level: str | None


class StudentDetailResponse(StudentResponse):
    """Detailed student response with aggregated data."""

    average_grade: float | None
    total_absences: int
    total_negative_events: int
    total_positive_events: int
    is_at_risk: bool
    performance_score: float | None = None


class GradeResponse(BaseModel):
    """Response model for grade data."""

    id: int
    subject: str
    teacher_name: str | None
    grade: float
    period: str


class AttendanceResponse(BaseModel):
    """Response model for attendance data."""

    id: int
    lessons_reported: int
    absence: int
    absence_justified: int
    late: int
    disturbance: int
    total_absences: int
    attendance: int
    total_negative_events: int
    total_positive_events: int
    period: str


class StudentListResponse(BaseModel):
    """Paginated list of students."""

    items: list[StudentDetailResponse]
    total: int
    page: int
    page_size: int


class ClassResponse(BaseModel):
    """Response model for class data."""

    id: UUID
    class_name: str
    grade_level: str
    student_count: int
    average_grade: float | None
    at_risk_count: int


class DashboardStats(BaseModel):
    """Dashboard KPI statistics."""

    total_students: int
    average_grade: float | None
    at_risk_count: int
    total_classes: int
    classes: list[ClassResponse]

File: client/src/lib/types/student.ts
import { z } from "zod";

export const StudentListItemSchema = z.object({
    student_tz: z.string(),
    student_name: z.string(),
    class_id: z.string(),
    class_name: z.string(),
    grade_level: z.string(),
    average_grade: z.number().nullable(),
    total_absences: z.number(),
    is_at_risk: z.boolean(),
});
export type StudentListItem = z.infer<typeof StudentListItemSchema>;

export const StudentListResponseSchema = z.object({
    items: z.array(StudentListItemSchema),
    total: z.number(),
    page: z.number(),
    page_size: z.number(),
});
export type StudentListResponse = z.infer<typeof StudentListResponseSchema>;

export const StudentDetailResponseSchema = z.object({
    student_tz: z.string(),
    student_name: z.string(),
    class_id: z.string(),
    class_name: z.string(),
    grade_level: z.string(),
    average_grade: z.number().nullable(),
    total_absences: z.number(),
    total_negative_events: z.number(),
    total_positive_events: z.number(),
    is_at_risk: z.boolean(),
    performance_score: z.number().nullable(),
});
export type StudentDetailResponse = z.infer<typeof StudentDetailResponseSchema>;

export const GradeResponseSchema = z.object({
    id: z.number(),
    subject: z.string(),
    teacher_name: z.string().nullable(),
    grade: z.number(),
    period: z.string(),
});
export type GradeResponse = z.infer<typeof GradeResponseSchema>;

export const AttendanceResponseSchema = z.object({
    id: z.number(),
    lessons_reported: z.number(),
    absence: z.number(),
    absence_justified: z.number(),
    late: z.number(),
    disturbance: z.number(),
    total_absences: z.number(),
    attendance: z.number(),
    total_negative_events: z.number(),
    total_positive_events: z.number(),
    period: z.string(),
});
export type AttendanceResponse = z.infer<typeof AttendanceResponseSchema>;

export const ClassResponseSchema = z.object({
    id: z.string(),
    class_name: z.string(),
    grade_level: z.string(),
    student_count: z.number(),
    average_grade: z.number().nullable(),
    at_risk_count: z.number(),
});
export type ClassResponse = z.infer<typeof ClassResponseSchema>;

export const DashboardStatsSchema = z.object({
    total_students: z.number(),
    average_grade: z.number().nullable(),
    at_risk_count: z.number(),
    total_classes: z.number(),
    classes: z.array(ClassResponseSchema),
});
export type DashboardStats = z.infer<typeof DashboardStatsSchema>;
